spharpy/spherical.py

```diff
@@ -1104,3 +1107,397 @@ def sph_identity_matrix(n_max, type='n-nm'):
+        """Get the gradient of the basis matrix with respect to theta."""
+        if self._basis_gradient_theta is None:
+            self._compute_basis_gradient()
+        return self._basis_gradient_theta
+
+    @property
+    def basis_gradient_phi(self):
+        """Get the gradient of the basis matrix with respect to phi."""  
+        if self._basis_gradient_phi is None:
+            self._compute_basis_gradient()
+        return self._basis_gradient_phi
+
+    def _compute_basis(self):
+        """
+        Compute the basis matrix for the SphericalHarmonics class.
+        """
+        if self.basis_type == "complex":
+            function = spherical_harmonic_basis
+        elif self.basis_type == "real":
+            function = spherical_harmonic_basis_real
+        else:
+            raise ValueError(
+                "Invalid basis type, should be either 'complex' or 'real'"
+            )
+        self._basis = function(
+            self.n_max, self.coordinates,
+            self.normalization, self.channel_convention)
+
+    def _compute_basis_gradient(self):
+        """
+        Compute the gradient of the basis matrix for the SphericalHarmonics class
+        """
+        if any(
+            (self.normalization in ["maxN", "sn3d"],
+             self.channel_convention == "fuma")
+        ):
+            raise ValueError(
+            f"Gradient computation not supported for normalization "
+            f"'{self.normalization}' and "
+            f"channel convention '{self.channel_convention}'."
+            )
+        else:
+            if self.basis_type == "complex":
+                function = spherical_harmonic_basis_gradient
+            elif self.basis_type == "real":
+                function = spherical_harmonic_basis_gradient_real
+            else:
+                raise ValueError(
+                    "Invalid basis type, should be either 'complex' or 'real'"
+                )
+            self._basis_gradient_theta, self._basis_gradient_phi = function(
+                self.n_max, self.coordinates)
+
+    @property
+    def basis_inv(self):
+        """Get or set the inverse basis matrix."""
+        if self._basis is None:
+            self._compute_basis()
+        if self._basis_inv is None:
+            self._compute_inverse()
+        return self._basis_inv
+
+    def _compute_inverse(self):
+        """
+        Compute the inverse basis matrix for the SphericalHarmonics class
+        """
+        if self._basis is None:
+            self._compute_basis()
+        _inv_flag = self.inverse_method
+        # warn when falling back to pseudo-inverse on a plain pf.Coordinates
+        if _inv_flag == "pseudo_inverse" and not isinstance(self.coordinates, SamplingSphere):
+            import warnings
+            warnings.warn(
+                "Coordinates is not a SamplingSphere, " \
+                "using Moore–Penrose pseudo-inverse instead of quadrature."
+            )
```

Comment:
```suggestion
```
not sure if this can happen, since we raise an error if we dont have a SamplingSphere and quadrature is set. So I think we can remove it here.

Reply:
If we want to keep it, we can import the warning at the top of the file instead